File: ios/runanywhere-sdks-latest/examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/presentation/vision/VLMViewModel.kt
package com.runanywhere.runanywhereai.presentation.vision

import android.Manifest
import android.app.Application
import android.content.pm.PackageManager
import android.net.Uri
import timber.log.Timber
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import androidx.camera.view.CameraController
import androidx.camera.view.LifecycleCameraController
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.viewModelScope
import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.extensions.VLM.VLMGenerationOptions
import com.runanywhere.sdk.public.extensions.VLM.VLMImage
import com.runanywhere.sdk.public.extensions.cancelVLMGeneration
import com.runanywhere.sdk.public.extensions.isVLMModelLoaded
import com.runanywhere.sdk.public.extensions.processImageStream
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.Executors

/**
 * UI state for VLM screen.
 * Mirrors iOS VLMViewModel published properties.
 */
data class VLMUiState(
    val isModelLoaded: Boolean = false,
    val loadedModelName: String? = null,
    val isProcessing: Boolean = false,
    val currentDescription: String = "",
    val error: String? = null,
    val selectedImageUri: Uri? = null,
    val showModelSelection: Boolean = false,
    val isCameraAuthorized: Boolean = false,
    val isAutoStreamingEnabled: Boolean = false,
)

/**
 * VLM ViewModel matching iOS VLMViewModel functionality.
 *
 * Manages:
 * - VLM model status
 * - Camera preview via CameraX LifecycleCameraController
 * - Frame capture via ImageAnalysis
 * - Image selection (from gallery)
 * - Image processing with streaming output
 * - Auto-streaming mode (live, every 2.5s)
 * - Generation cancellation
 *
 * iOS Reference: examples/ios/RunAnywhereAI/.../Features/Vision/VLMViewModel.swift
 */
class VLMViewModel(application: Application) : AndroidViewModel(application) {
    companion object {
        private const val AUTO_STREAM_INTERVAL_MS = 2500L
    }

    private val _uiState = MutableStateFlow(VLMUiState())
    val uiState: StateFlow<VLMUiState> = _uiState.asStateFlow()

    private var generationJob: Job? = null
    private var autoStreamJob: Job? = null

    // Camera — uses LifecycleCameraController for simplicity
    private var cameraController: LifecycleCameraController? = null
    private var currentFrameRgb: ByteArray? = null
    private var currentFrameWidth: Int = 0
    private var currentFrameHeight: Int = 0
    private val frameLock = Any()

    // Background executor for frame analysis (avoid blocking main thread)
    private val analysisExecutor = Executors.newSingleThreadExecutor()

    init {
        viewModelScope.launch {
            checkModelStatus()
            checkCameraPermission()
        }
    }

    // MODEL

    fun checkModelStatus() {
        try {
            val isLoaded = RunAnywhere.isVLMModelLoaded
            _uiState.update { it.copy(isModelLoaded = isLoaded) }
            Timber.d("VLM model loaded: $isLoaded")
        } catch (e: Exception) {
            Timber.e(e, "Failed to check VLM model status: ${e.message}")
            _uiState.update { it.copy(isModelLoaded = false) }
        }
    }

    fun onModelLoaded(modelName: String) {
        _uiState.update {
            it.copy(
                isModelLoaded = true,
                loadedModelName = modelName,
                showModelSelection = false,
            )
        }
    }

    fun setShowModelSelection(show: Boolean) {
        _uiState.update { it.copy(showModelSelection = show) }
    }

    // CAMERA - Mirrors iOS setupCamera / startCamera / stopCamera
    // Uses LifecycleCameraController (CameraX recommended API)

    fun checkCameraPermission() {
        val context = getApplication<Application>()
        val granted = ContextCompat.checkSelfPermission(
            context, Manifest.permission.CAMERA,
        ) == PackageManager.PERMISSION_GRANTED
        _uiState.update { it.copy(isCameraAuthorized = granted) }
    }

    fun onCameraPermissionResult(granted: Boolean) {
        _uiState.update { it.copy(isCameraAuthorized = granted) }
    }

    /**
     * Bind camera preview + image analysis to the given PreviewView.
     * Mirrors iOS setupCamera() + startCamera().
     *
     * Uses LifecycleCameraController which automatically manages Preview + ImageAnalysis.
     */
    fun bindCamera(previewView: PreviewView, lifecycleOwner: LifecycleOwner) {
        if (cameraController != null) return

        val context = getApplication<Application>()

        val controller = LifecycleCameraController(context).apply {
            cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA
            setEnabledUseCases(CameraController.IMAGE_ANALYSIS)
            imageAnalysisBackpressureStrategy = ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST
            imageAnalysisOutputImageFormat = ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888

            setImageAnalysisAnalyzer(analysisExecutor) { imageProxy ->
                captureFrame(imageProxy)
            }
        }

        controller.bindToLifecycle(lifecycleOwner)
        previewView.controller = controller
        cameraController = controller
    }

    fun unbindCamera() {
        cameraController?.unbind()
        cameraController = null
    }

    /**
     * Capture the latest camera frame as RGB bytes.
     * Mirrors iOS captureOutput delegate that stores currentFrame (CVPixelBuffer).
     *
     * CameraX delivers RGBA_8888 format; we strip alpha to get RGB.
     */
    private fun captureFrame(imageProxy: ImageProxy) {
        try {
            val plane = imageProxy.planes[0]
            val buffer = plane.buffer
            val width = imageProxy.width
            val height = imageProxy.height
            val rowStride = plane.rowStride
            val pixelStride = plane.pixelStride

            // RGBA_8888 -> RGB (strip alpha), handle row stride padding
            val rgbSize = width * height * 3
            val rgb = ByteArray(rgbSize)
            var rgbIdx = 0
            for (row in 0 until height) {
                for (col in 0 until width) {
                    val srcIdx = row * rowStride + col * pixelStride
                    if (srcIdx + 2 < buffer.limit()) {
                        rgb[rgbIdx++] = buffer[srcIdx]       // R
                        rgb[rgbIdx++] = buffer[srcIdx + 1]   // G
                        rgb[rgbIdx++] = buffer[srcIdx + 2]   // B
                    } else {
                        rgbIdx += 3 // skip pixel but keep alignment
                    }
                }
            }

            synchronized(frameLock) {
                currentFrameRgb = rgb
                currentFrameWidth = width
                currentFrameHeight = height
            }
        } catch (e: Exception) {
            Timber.e("Frame capture failed: ${e.message}")
        } finally {
            imageProxy.close()
        }
    }

    // DESCRIBE - Mirrors iOS describeCurrentFrame / describeImage

    /**
     * Describe the current camera frame. Mirrors iOS describeCurrentFrame().
     */
    fun describeCurrentFrame() {
        if (!_uiState.value.isModelLoaded) {
            _uiState.update { it.copy(error = "Please load a model first") }
            return
        }

        val frameData: ByteArray
        val w: Int
        val h: Int

        synchronized(frameLock) {
            frameData = currentFrameRgb ?: run {
                _uiState.update { it.copy(error = "No camera frame available") }
                return
            }
            w = currentFrameWidth
            h = currentFrameHeight
        }

        if (_uiState.value.isProcessing) return

        generationJob?.cancel()
        _uiState.update {
            it.copy(isProcessing = true, currentDescription = "", error = null)
        }

        generationJob = viewModelScope.launch {
            try {
                val image = VLMImage.fromRGBPixels(frameData, w, h)
                val options = VLMGenerationOptions(maxTokens = 200, temperature = 0.7f)

                Timber.i("Describing current camera frame (${w}x${h})")

                RunAnywhere.processImageStream(
                    image,
                    "Describe what you see briefly.",
                    options,
                ).collect { token ->
                    _uiState.update { it.copy(currentDescription = it.currentDescription + token) }
                }

                _uiState.update { it.copy(currentDescription = it.currentDescription.trim()) }
                Timber.i("Frame description completed")
            } catch (e: Exception) {
                Timber.e(e, "Frame description failed: ${e.message}")
                _uiState.update { it.copy(error = "Processing failed: ${e.message}") }
            } finally {
                _uiState.update { it.copy(isProcessing = false) }
            }
        }
    }

    /**
     * Process a gallery image. Mirrors iOS describeImage(uiImage:).
     */
    fun processSelectedImage(prompt: String = "Describe this image in detail.") {
        val uri = _uiState.value.selectedImageUri ?: return

        if (!_uiState.value.isModelLoaded) {
            _uiState.update { it.copy(error = "No VLM model loaded. Please select a model first.") }
            return
        }

        generationJob?.cancel()
        cancelGeneration()

        _uiState.update {
            it.copy(isProcessing = true, currentDescription = "", error = null)
        }

        generationJob = viewModelScope.launch {
            var tempFile: File? = null
            try {
                tempFile = copyUriToTempFile(uri) ?: throw Exception("Failed to read image")
                val image = VLMImage.fromFilePath(tempFile.absolutePath)
                val options = VLMGenerationOptions(maxTokens = 300, temperature = 0.7f)

                Timber.i("Starting VLM streaming for image: ${tempFile.name}")

                RunAnywhere.processImageStream(image, prompt, options)
                    .collect { token ->
                        _uiState.update { it.copy(currentDescription = it.currentDescription + token) }
                    }

                _uiState.update { it.copy(currentDescription = it.currentDescription.trim()) }
                Timber.i("VLM streaming completed")
            } catch (e: Exception) {
                Timber.e(e, "VLM processing failed: ${e.message}")
                _uiState.update { it.copy(error = "Processing failed: ${e.message}") }
            } finally {
                tempFile?.delete()
                _uiState.update { it.copy(isProcessing = false) }
            }
        }
    }

    // AUTO-STREAMING - Mirrors iOS toggleAutoStreaming / startAutoStreaming

    fun toggleAutoStreaming() {
        if (_uiState.value.isAutoStreamingEnabled) {
            stopAutoStreaming()
        } else {
            startAutoStreaming()
        }
    }

    private fun startAutoStreaming() {
        if (autoStreamJob != null) return
        _uiState.update { it.copy(isAutoStreamingEnabled = true) }

        autoStreamJob = viewModelScope.launch {
            while (_uiState.value.isAutoStreamingEnabled) {
                // Wait for any current processing to finish
                while (_uiState.value.isProcessing) {
                    delay(100)
                    if (!_uiState.value.isAutoStreamingEnabled) return@launch
                }

                describeCurrentFrameForAutoStream()

                delay(AUTO_STREAM_INTERVAL_MS)
            }
        }
    }

    fun stopAutoStreaming() {
        autoStreamJob?.cancel()
        autoStreamJob = null
        _uiState.update { it.copy(isAutoStreamingEnabled = false) }
    }

    /**
     * Auto-stream variant — smoother transition, shorter prompt, errors logged only.
     * Mirrors iOS describeCurrentFrameForAutoStream().
     */
    private suspend fun describeCurrentFrameForAutoStream() {
        val frameData: ByteArray
        val w: Int
        val h: Int

        synchronized(frameLock) {
            frameData = currentFrameRgb ?: return
            w = currentFrameWidth
            h = currentFrameHeight
        }

        if (_uiState.value.isProcessing) return

        _uiState.update { it.copy(isProcessing = true, error = null) }

        var newDescription = ""
        try {
            val image = VLMImage.fromRGBPixels(frameData, w, h)
            val options = VLMGenerationOptions(maxTokens = 100, temperature = 0.7f)

            RunAnywhere.processImageStream(
                image,
                "Describe what you see in one sentence.",
                options,
            ).collect { token ->
                newDescription += token
                _uiState.update { it.copy(currentDescription = newDescription) }
            }
            _uiState.update { it.copy(currentDescription = newDescription.trim()) }
        } catch (e: Exception) {
            Timber.e("Auto-stream VLM error: ${e.message}")
        } finally {
            _uiState.update { it.copy(isProcessing = false) }
        }
    }

    // CANCEL

    fun cancelGeneration() {
        try {
            RunAnywhere.cancelVLMGeneration()
            generationJob?.cancel()
            _uiState.update { it.copy(isProcessing = false) }
            Timber.d("VLM generation cancelled")
        } catch (e: Exception) {
            Timber.e(e, "Failed to cancel VLM generation: ${e.message}")
        }
    }

    // IMAGE SELECTION

    fun setSelectedImage(uri: Uri?) {
        _uiState.update {
            it.copy(selectedImageUri = uri, currentDescription = "", error = null)
        }
    }

    // HELPERS

    private suspend fun copyUriToTempFile(uri: Uri): File? = withContext(Dispatchers.IO) {
        try {
            val context = getApplication<Application>()
            val inputStream = context.contentResolver.openInputStream(uri) ?: return@withContext null
            val tempFile = File.createTempFile("vlm_image_", ".jpg", context.cacheDir)
            inputStream.use { input ->
                FileOutputStream(tempFile).use { output ->
                    input.copyTo(output)
                }
            }
            tempFile
        } catch (e: Exception) {
            Timber.e(e, "Failed to copy URI to temp file: ${e.message}")
            null
        }
    }

    override fun onCleared() {
        super.onCleared()
        generationJob?.cancel()
        autoStreamJob?.cancel()
        unbindCamera()
        analysisExecutor.shutdown()
    }
}

File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/features/vision/vlm_view_model.dart
import 'dart:async';

import 'package:camera/camera.dart';
import 'package:flutter/material.dart';
import 'package:runanywhere/runanywhere.dart' as sdk;
import 'package:runanywhere_ai/core/services/permission_service.dart';

/// VLMViewModel - State management for VLM camera view
///
/// Mirrors iOS VLMViewModel.swift exactly:
/// - Camera management (authorization, initialization, disposal)
/// - Model status tracking (loaded state, model name)
/// - Single capture mode (camera frame → description)
/// - Gallery photo mode (picked image → detailed description)
/// - Auto-streaming mode (live 2.5s interval captures)
/// - Token-by-token streaming display
/// - Error handling and cancellation
class VLMViewModel extends ChangeNotifier {
  // MARK: - State Properties

  bool _isModelLoaded = false;
  String? _loadedModelName;
  bool _isProcessing = false;
  String _currentDescription = '';
  String? _error;
  bool _isCameraAuthorized = false;
  bool _isCameraInitialized = false;
  bool _isAutoStreamingEnabled = false;

  // Getters
  bool get isModelLoaded => _isModelLoaded;
  String? get loadedModelName => _loadedModelName;
  bool get isProcessing => _isProcessing;
  String get currentDescription => _currentDescription;
  String? get error => _error;
  bool get isCameraAuthorized => _isCameraAuthorized;
  bool get isCameraInitialized => _isCameraInitialized;
  bool get isAutoStreamingEnabled => _isAutoStreamingEnabled;

  // MARK: - Camera Management

  CameraController? _cameraController;
  CameraController? get cameraController => _cameraController;

  Timer? _autoStreamTimer;
  static const autoStreamInterval = Duration(seconds: 2, milliseconds: 500);

  // MARK: - Camera Initialization

  /// Initialize camera with back camera (or first available)
  /// Request BGRA format (preferred for iOS, Android may fallback to YUV)
  Future<void> initializeCamera() async {
    try {
      final cameras = await availableCameras();
      if (cameras.isEmpty) {
        debugPrint('❌ No cameras available');
        return;
      }

      // Select back camera (or first available)
      final camera = cameras.firstWhere(
        (c) => c.lensDirection == CameraLensDirection.back,
        orElse: () => cameras.first,
      );

      // Create controller with BGRA format request (iOS preferred, Android fallback to YUV)
      _cameraController = CameraController(
        camera,
        ResolutionPreset.medium,
        imageFormatGroup: ImageFormatGroup.bgra8888,
      );

      await _cameraController!.initialize();

      _isCameraInitialized = true;
      notifyListeners();

      debugPrint('✅ Camera initialized: ${camera.lensDirection}');
    } catch (e) {
      debugPrint('❌ Camera initialization failed: $e');
      _error = 'Failed to initialize camera: $e';
      notifyListeners();
    }
  }

  /// Dispose camera controller
  void disposeCamera() {
    unawaited(_cameraController?.dispose());
    _cameraController = null;
    _isCameraInitialized = false;
    notifyListeners();
  }

  /// Check and request camera permission
  Future<void> checkCameraAuthorization(BuildContext context) async {
    _isCameraAuthorized =
        await PermissionService.shared.requestCameraPermission(context);
    notifyListeners();
  }

  // MARK: - Model Management

  /// Check if VLM model is loaded
  Future<void> checkModelStatus() async {
    _isModelLoaded = sdk.RunAnywhere.isVLMModelLoaded;
    if (_isModelLoaded) {
      _loadedModelName = sdk.RunAnywhere.currentVLMModelId;
    } else {
      _loadedModelName = null;
    }
    notifyListeners();
  }

  /// Handle model selection from sheet
  /// Takes the app's ModelInfo and loads the SDK model by ID
  Future<void> onModelSelected(
      String modelId, String modelName, BuildContext context) async {
    try {
      debugPrint('🎯 Loading VLM model: $modelId');
      await sdk.RunAnywhere.loadVLMModel(modelId);
      _isModelLoaded = true;
      _loadedModelName = modelName;
      notifyListeners();
      debugPrint('✅ VLM model loaded: $modelName');
    } catch (e) {
      debugPrint('❌ Failed to load VLM model: $e');
      _error = 'Failed to load model: $e';
      notifyListeners();
      if (context.mounted) {
        unawaited(
          ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(content: Text('Failed to load model: $e')),
          ).closed.then((_) => null),
        );
      }
    }
  }

  // MARK: - Image Processing - Single Capture

  /// Describe the current camera frame (single capture mode)
  /// Matches iOS describeCurrentFrame()
  Future<void> describeCurrentFrame() async {
    if (_isProcessing || !_isCameraInitialized || _cameraController == null) {
      return;
    }

    _isProcessing = true;
    _error = null;
    _currentDescription = '';
    notifyListeners();

    try {
      // Capture image from camera
      final xFile = await _cameraController!.takePicture();

      // Create VLMImage from file path
      final image = sdk.VLMImage.filePath(xFile.path);

      // Process image with streaming
      final result = await sdk.RunAnywhere.processImageStream(
        image,
        prompt: 'Describe what you see briefly.',
        maxTokens: 200,
      );

      // Listen to stream and append tokens
      final buffer = StringBuffer(_currentDescription);
      await for (final token in result.stream) {
        buffer.write(token);
        _currentDescription = buffer.toString();
        notifyListeners();
      }

      debugPrint('✅ Single capture complete: ${_currentDescription.length} chars');
    } catch (e) {
      debugPrint('❌ Single capture error: $e');
      _error = e.toString();
      notifyListeners();
    } finally {
      _isProcessing = false;
      notifyListeners();
    }
  }

  // MARK: - Image Processing - Gallery Photo

  /// Describe a picked image from gallery
  /// Matches iOS describeImage(_:)
  Future<void> describePickedImage(String imagePath) async {
    _isProcessing = true;
    _error = null;
    _currentDescription = '';
    notifyListeners();

    try {
      // Create VLMImage from file path
      final image = sdk.VLMImage.filePath(imagePath);

      // Process image with streaming (more detailed prompt)
      final result = await sdk.RunAnywhere.processImageStream(
        image,
        prompt: 'Describe this image in detail.',
        maxTokens: 300,
      );

      // Listen to stream and append tokens
      final buffer = StringBuffer(_currentDescription);
      await for (final token in result.stream) {
        buffer.write(token);
        _currentDescription = buffer.toString();
        notifyListeners();
      }

      debugPrint('✅ Gallery photo described: ${_currentDescription.length} chars');
    } catch (e) {
      debugPrint('❌ Gallery photo error: $e');
      _error = e.toString();
      notifyListeners();
    } finally {
      _isProcessing = false;
      notifyListeners();
    }
  }

  // MARK: - Auto-Streaming (Live Mode)

  /// Toggle auto-streaming mode
  /// Matches iOS toggleAutoStreaming()
  void toggleAutoStreaming() {
    _isAutoStreamingEnabled = !_isAutoStreamingEnabled;
    notifyListeners();

    if (_isAutoStreamingEnabled) {
      _startAutoStreaming();
    } else {
      stopAutoStreaming();
    }
  }

  /// Start auto-streaming with periodic timer
  void _startAutoStreaming() {
    _autoStreamTimer?.cancel();
    _autoStreamTimer = Timer.periodic(autoStreamInterval, (timer) {
      if (!_isProcessing) {
        unawaited(_describeCurrentFrameForAutoStream());
      }
    });
    debugPrint('🔴 Auto-streaming started (${autoStreamInterval.inMilliseconds}ms interval)');
  }

  /// Stop auto-streaming
  void stopAutoStreaming() {
    _autoStreamTimer?.cancel();
    _autoStreamTimer = null;
    _isAutoStreamingEnabled = false;
    notifyListeners();
    debugPrint('⏹️ Auto-streaming stopped');
  }

  /// Describe current frame for auto-stream (live mode)
  /// Matches iOS describeCurrentFrameForAutoStream()
  /// - Shorter prompt for quick responses
  /// - Don't clear description (smooth transition)
  /// - Errors only logged, not shown to user
  Future<void> _describeCurrentFrameForAutoStream() async {
    if (_isProcessing || !_isCameraInitialized || _cameraController == null) {
      return;
    }

    _isProcessing = true;
    notifyListeners();

    // Build new description in local var (per iOS pattern)
    String newDescription = '';

    try {
      // Capture image from camera
      final xFile = await _cameraController!.takePicture();

      // Create VLMImage from file path
      final image = sdk.VLMImage.filePath(xFile.path);

      // Process image with streaming (shorter prompt for live mode)
      final result = await sdk.RunAnywhere.processImageStream(
        image,
        prompt: 'Describe what you see in one sentence.',
        maxTokens: 100,
      );

      // Listen to stream and build description
      final buffer = StringBuffer(newDescription);
      await for (final token in result.stream) {
        buffer.write(token);
        newDescription = buffer.toString();
        _currentDescription = newDescription;
        notifyListeners();
      }

      debugPrint('🔴 Auto-stream capture complete: ${newDescription.length} chars');
    } catch (e) {
      // Only log errors in auto-stream mode (per iOS pattern)
      debugPrint('⚠️ Auto-stream error (non-critical): $e');
      // Don't set _error in auto-stream mode
    } finally {
      _isProcessing = false;
      notifyListeners();
    }
  }

  // MARK: - Cancellation

  /// Cancel ongoing VLM generation
  Future<void> cancelGeneration() async {
    unawaited(sdk.RunAnywhere.cancelVLMGeneration());
    debugPrint('🛑 VLM generation cancelled');
  }

  // MARK: - Cleanup

  @override
  void dispose() {
    _autoStreamTimer?.cancel();
    unawaited(_cameraController?.dispose());
    super.dispose();
  }
}
